### ISSUE TITLE
Global session variable when used in rails

### ISSUE THREAD

**Issue by chrismanderson** [adelevie/parse-ruby-client#152]
Hello! Ran into an issue when using the gem in a rails app - it seems that the session_token set on the client is global across all rails sessions, so if I am logged in with user A, open another browse, login with user B, go back to A, and attempt to use this gem, the parse query connects via user B instead of user A. Does there happen to be a straightforward way of avoiding this?

**Comment by adelevie** [comment]
Hmm... that doesn't sound good. Could you provide a quick code sample/synopsis that triggers this bug? Thanks for reporting.

**Comment by chrismanderson** [comment]
Preparing a larger sample, but the basic gist is - 

in `config/initializers/application.rb` we initializer `Parse`.

    Parse.init :application_id => ENV["parse_application_id"],
           :api_key       => ENV["parse_api_key"],
           :quiet           => false     

Then, have a simple sessions_controller to do login.

	def create
		user = Parse::User.authenticate(params[:email], params[:password])
		if user
			session[:user_id] = user.id
			redirect_to root_path
		else
			flash.now.alert = "Invalid email or password"
			render "new"
		end
		rescue Parse::ParseProtocolError => e
			redirect_to new_session_path, :flash => { :error => "Unable to Login. Please try again." }
	end

Lastly, have an index page of sorts.

    class MenuController < ApplicationController
		def index
			meal_query = Parse::Query.new("Meal")
			meal_query.eq("availableDate", Parse::Date.new(DateTime.now.beginning_of_day.change(:offset => 'EST')))
    		@meals = meal_query.get
    		flash.now[:notice] = "We're closed right now, please check back" unless store_status
		end

		def store_status
	  		@store_status = false
	  
		    if( session[:store_status] ) 
		      @store_status = session[:store_status]["is_store_open"]
		    else
		      cloudStatus = Parse::Cloud::Function.new("storeStatus").call()
		      session[:store_status] = cloudStatus
		      @store_status = cloudStatus["is_store_open"]
	    	end
	  
	   		@store_status = true
      	end
    end

If I login with user A in one browser, login with user B in another browser, when I go back to browser A and refresh that index page, the `storeStatus` cloud function is called with the user B session token.

**Comment by adelevie** [comment]
https://github.com/adelevie/parse-ruby-client/blob/master/lib/parse/user.rb#L17 looks like the offending line. This speaks to an early design decision by the original author (not me) to rely heavily on the `Parse` singleton. Do you think there's a simple way to fix this or would there need to be a larger overhaul?

Thanks again for reporting and I'm surprised this wasn't picked up sooner. I really don't have much time for this gem anymore (sadly), but I'll try and help along the way as much as I can.

**Comment by adelevie** [comment]
Maybe use this as a starting point for a PR:

```ruby
    def self.get_session_token(username, password)  # might want to rethink the name
      body = {
        "username" => username,
        "password" => password
      }

      Parse.client.request(Parse::Protocol::USER_LOGIN_URI, :get, nil, body)
    end
```

Now all Parse query methods will need an optional argument/parameter to take this session token. I think. Does this approach make sense? Are there any other ways around this?

**Comment by adelevie** [comment]
Does this help in any way?

```ruby
class ParseWrapper
  attr_reader :parse

  def initialize
    @parse = Parse.dup
  end
end

parse = ParseWrapper.new.parse
```

**Comment by adelevie** [comment]
pinging @ericcj amd @bcherry who might have some insight.

**Comment by chrismanderson** [comment]
Yeah that last comment might be useful, will take a look, thanks.

**Comment by ericcj** [comment]
The only way to avoid it with the current code is to clear that variable
after each request and re auth at the start and/or cache them in the rails
session.  See the multithreading issues for some discussion of of making
the client an optional param to the query and object initializers and/or
give it builder methods like client.query instead of Query.new

**Comment by rhymes** [comment]
Can you guys please take a look at #165 ?

**Comment by Xavdidtheshadow** [comment]
@chrismanderson is this fixed by the recently merged #165? You can specify github in your gemfile to use it before it gets pushed to rubygems.

**Comment by Xavdidtheshadow** [comment]
I'll assume this is fixed. Reopen if not.
